### PR TITLE
Wip/drothe 9209 sub device area handshake

### PR DESCRIPTION
**Pull request by d-rothe**

### CODE REVIEW

File: device_backends/Subdevice/src/SubdeviceBackend.cc [hunk 1]
```diff
@@ -351,7 +343,7 @@ namespace ChimeraTK {
   template<typename UserType>
   boost::shared_ptr<NDRegisterAccessor<UserType>> SubdeviceBackend::getRegisterAccessor_area(
       const RegisterPath& registerPathName, size_t numberOfWords, size_t wordOffsetInRegister, AccessModeFlags flags) {
-    assert(type == Type::area);
+    assert(hasAreaParam());
```
Comment (mhier) [inline device_backends/Subdevice/src/SubdeviceBackend.cc:346]: I think this must stay at the old assert, since this case is only for the area type and not for the areaHandshake, right?

File: device_backends/Subdevice/src/SubdeviceBackend.cc [hunk 2]
```diff
@@ -437,7 +443,7 @@ namespace ChimeraTK {
   template<>
   boost::shared_ptr<NDRegisterAccessor<int32_t>> SubdeviceBackend::getRegisterAccessor_area<int32_t>(
       const RegisterPath& registerPathName, size_t numberOfWords, size_t wordOffsetInRegister, AccessModeFlags flags) {
-    assert(type == Type::area);
+    assert(hasAreaParam());
```
Comment (mhier) [inline device_backends/Subdevice/src/SubdeviceBackend.cc:446]: see above...

File: device_backends/Subdevice/src/SubdeviceRegisterAccessor.cc [hunk 3]
```diff
@@ -26,41 +26,51 @@ namespace ChimeraTK {
...
         }
-        _accAddress->accessData(0) = adr;
-        _accAddress->write();
-        usleep(_backend->addressToDataDelay);
-        for(size_t innerOffset = 0; innerOffset < _accData->getNumberOfSamples(); ++innerOffset) {
-          if(idx < _numberOfWords) {
-            _accData->accessData(0, innerOffset) = _buffer[idx];
-          }
-          else {
+        if(_backend->hasAreaParam()) {
```
Comment (mhier) [inline device_backends/Subdevice/src/SubdeviceRegisterAccessor.cc:57]: Even though this seems to have the right behaviour, I would check for the areaHandshake type instead of calling this function. The code in the if case is not executed for the normal area type, hence this might be confusing.

File: device_backends/Subdevice/src/SubdeviceRegisterAccessor.cc [hunk 4]
```diff
@@ -26,41 +26,51 @@ namespace ChimeraTK {
...
+          ++idx;
+        }
+        else {
+          _accAddress->accessData(0) = adr;
+          _accAddress->write();
+          usleep(_backend->addressToDataDelay);
+          for(size_t innerOffset = 0; innerOffset < _accDataArea->getNumberOfSamples(); ++innerOffset) {
             // pad data with zeros, if _numberOfWords isn't an integer multiple of _accData->getNumberOfSamples()
-            _accData->accessData(0, innerOffset) = 0;
+            int32_t val = (idx < _numberOfWords) ? _buffer[idx] : 0;
```
Comment (mhier) [inline device_backends/Subdevice/src/SubdeviceRegisterAccessor.cc:68]: I like this change, but please next time do it in a separate PR (or even commit it right away to the master). It is unrelated to the actual new feature...

File: tests/executables_src/testSubdeviceBackend.cpp [hunk 5]
```diff
@@ -586,6 +586,63 @@ BOOST_AUTO_TEST_CASE(test3regsByteOffset1) {
...
+  Device dev;
+  dev.open("SUBDEV4");
+  Device target;
+  target.open("TARGET1");
+
+  auto acc1 = dev.getScalarRegisterAccessor<double>("APP.0.MY_REGISTER1");
+  auto acc2 = dev.getScalarRegisterAccessor<double>("APP.0.MY_REGISTER2");
+  auto acc3 = dev.getOneDRegisterAccessor<int>("APP.0.MY_AREA1", 6, 0);
+  auto accArea = target.getOneDRegisterAccessor<int32_t>("APP.0.THE_AREA", 10, 0, {AccessMode::raw});
+  target.getScalarRegisterAccessor<int32_t>("APP.0.THE_AREA");
```
Comment (mhier) [inline tests/executables_src/testSubdeviceBackend.cpp:601]: What is the purpose of this line?

File: tests/executables_src/testSubdeviceBackend.cpp [hunk 6]
```diff
@@ -586,6 +586,63 @@ BOOST_AUTO_TEST_CASE(test3regsByteOffset1) {
...
+  // the dummyForAreaHandshake backend which we use for this test does not set back the status register. we do it manually from the test, and count how often we need do so
+  // Like this we can check that the accessor waits on status==0 _each_ time before writing,
+  // in particular each array entry counts.
+  while(!done) {
+    countStatusResets++;
+    accS = 0;
+    accS.write();
+    usleep(200000);
+  }
+  BOOST_CHECK(countStatusResets >= 8); // do not strictly compare with 8 since we have little control over timing
```
Comment (mhier) [inline tests/executables_src/testSubdeviceBackend.cpp:633]: Why no exact comparison? The accessor should always write exactly all elements once. How can it become bigger than 8? The test is writing in total 8 elements exactly. All writes are complete before this comparison is done (this is  synchronised via `done`). This seems to be somewhat fishy...

**Review by mhier — CHANGES_REQUESTED** [review]

File: device_backends/Subdevice/include/SubdeviceBackend.h [hunk 7]
```diff
@@ -136,6 +143,13 @@ namespace ChimeraTK {
     void setException() override;
 
     void activateAsyncRead() noexcept override;
+
+    bool hasAreaParam() { return type == Type::area || type == Type::areaHandshake; }
```
[the remark this answers is not in the document]
Comment (d-rothe) [reply device_backends/Subdevice/include/SubdeviceBackend.h]: changed.

File: device_backends/Subdevice/src/SubdeviceBackend.cc (hunk 1 above)
Comment (d-rothe) [reply device_backends/Subdevice/src/SubdeviceBackend.cc]: yes

File: tests/executables_src/testSubdeviceBackend.cpp (hunk 5 above)
Comment (d-rothe) [reply tests/executables_src/testSubdeviceBackend.cpp]: none. deleted.

File: tests/executables_src/testSubdeviceBackend.cpp (hunk 6 above)
Comment (d-rothe) [reply tests/executables_src/testSubdeviceBackend.cpp]: changed to waiting for status=busy in loop
